codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/requestcompression/RequestCompression.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.requestcompression;
+
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+
+import java.util.List;
+import java.util.Set;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.GoDelegator;
+import software.amazon.smithy.go.codegen.GoSettings;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.SmithyGoDependency;
+import software.amazon.smithy.go.codegen.SymbolUtils;
+import software.amazon.smithy.go.codegen.integration.ConfigField;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.traits.RequestCompressionTrait;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.MapUtils;
+
+
+public final class RequestCompression implements GoIntegration {
+    private static final String ADD_REQUEST_COMPRESSION = "addRequestCompression";
+
+    private static final String ADD_REQUEST_COMPRESSION_INTERNAL = "AddRequestCompression";
+
+    private static final String DISABLE_REQUEST_COMPRESSION = "DisableRequestCompression";
+
+    private static final String REQUEST_MIN_COMPRESSION_SIZE_BYTES = "RequestMinCompressSizeBytes";
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        ServiceShape service = settings.getService(model);
+        if (!isRequestCompressionService(model, service)) {
+            return;
+        }
+
+        Set<String> algorithms = RequestCompressionTrait.SUPPORTED_COMPRESSION_ALGORITHMS;
+        goDelegator.useShapeWriter(service, writeMiddlewareHelper(algorithms));
```

Comment:
We have to generate this *per-operation* since the compression trait is applied there.

e.g. consider the following three modeled operations and their expected codegen:

```
@requestCompression{
    encodings: ["gzip"]
}
operation Foo {
  // ...
}
// should generate:
func addOperationFooRequestCompressionMiddleware(...) {
	return stack.Serialize.Add(&requestCompression{
                // ...
		compressAlgorithms: []string{"gzip"},
	}, middleware.After)
}
// and the operation stack setup should now include:
func (c *Client) addOperationFooMiddlewares(...) {
    // ...
    if err = addOperationFooRequestCompressionMiddleware(...); err != nil {
        return err
    }
    // ...
}

@requestCompression{
    encodings: ["br", "gzip"]
}
operation Bar {
  // ...
}
// should generate:
func addOperationBarRequestCompressionMiddleware(...) {
	return stack.Serialize.Add(&requestCompression{
                // ...
		compressAlgorithms: []string{"br", "gzip"},
	}, middleware.After)
}
// and the operation stack setup should now include:
func (c *Client) addOperationBarMiddlewares(...) {
    // ...
    if err = addOperationBarRequestCompressionMiddleware(...); err != nil {
        return err
    }
    // ...
}

// no request compression
operation Bar {
  // ...
}
// should not generate a middleware, and the operation stack setup shouldn't have anything new
```

See `AwsHttpChecksumGenerator.java` for an example of how this is done.

Reply:
Okay I previously thought the `SUPPORTED_COMPRESSION_ALGORITHMS` of that trait is static so service level middleware is enough, but seems like we have multi-encodings for diff ops, I will change that